internal/logging/logging.go

```diff
@@ -187,3 +187,35 @@ func isValidLogLevel(level string) bool {
 
 	return false
 }
+
+// PluginOutputMonitor creates an io.Writer that will warn about any writes in
+// the default logger. This is used to catch unexpected output from plugins,
+// notifying them about the problem as well as surfacing the lost data for
+// context.
+func PluginOutputMonitor(source string) io.Writer {
+	return pluginOutputMonitor{
+		source: source,
+		log:    logger,
+	}
+}
+
+// pluginOutputMonitor is an io.Writer that logs all writes as
+// "unexpected data" with the source name.
+type pluginOutputMonitor struct {
+	source string
+	log    hclog.Logger
+}
+
+func (w pluginOutputMonitor) Write(d []byte) (int, error) {
+	// Limit the write size to 1024 bytes We're not expecting any data to come
+	// through this channel, so accidental writes will usually be stray fmt
+	// debugging statements and the like, but we want to provide context to the
+	// provider to indicate what the unexpected data might be.
+	n := len(d)
+	if n > 1024 {
+		d = append(d[:1024], '.', '.', '.')
+	}
+
+	w.log.Warn("unexpected data", w.source, strings.TrimSpace(string(d)))
```

Comment:
@jbardin I'm late, but how do we feel about pointing people to terraform-plugin-log in the warning message here? I can draft something and PR it if this uncontroversial.

Reply:
I think output here is most likely going to be accidental, and the solution is to remove the accidental print statement rather than reconsidering how logging works. A much smaller issue is that `terraform-plugin-log` just may not be relevant for some providers. If you can craft a useful message to put in here, I don't think it would hurt.

We can't really add much more text to the log line without being overly verbose, but we could gate a single larger message to print on the first write.

Reply:
I think the use case I'm trying to optimize for is "first time provider developer reaches for fmt.Println as a logging solution and doesn't even know about terraform-plugin-log, and so when they get 'unexpected data' that doesn't really tell them what to do to make their logging work."

I think anyone that sees these messages is gonna need to do some unintuitive work to get their logs working; they either need to use `terraform-plugin-log` or hoist stderr before serving.

I think, ideally, we'd point to https://www.terraform.io/docs/extend/debugging.html or something, the more I think about this, so we can give provider developers the nuance they're looking for on how to write log output that will function appropriately.

Reply:
```suggestion
	w.log.Warn("unexpected data (see https://www.terraform.io/docs/extend/debugging.html)", w.source, strings.TrimSpace(string(d)))
```